rfcs/convergence/props-typing-improvements.md

```diff
@@ -0,0 +1,164 @@
+# RFC: Improved type checking for component props
+
+Contributors: @behowell
+
+## Summary
+
+This RFC seeks to improve type checking for slot props and props with default values in converged components. It'll make it easier for component authors to "do the right thing" when definining the State type for the component, which reduces the chances for bugs and makes future maintenance easier.
+
+## Problem statement
+
+1. The slot props for a component are mentioned in several different locations in code, without any explicit link between the names and types, other than that they happen to mention the same props.
+2. Some component props are given default values in the `useComponent` hook, but that is not represented in the type system (neither when the default values are provided, nor when they are used in, say `useComponentStyles`).
+
+For example, given a component named `Widget` with slots `icon` and `text`, and a prop `propWithDefaultValue` that is given a default value in `useWidget`:
+
+```typescript
+export interface WidgetProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  icon?: ShorthandProps<IconProps>;
+  text?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;
+
+  // This prop has a default defined in useWidget:
+  propWithDefaultValue?: string;
+
+  // This prop does not have a default value in useWidget (undefined by default)
+  anotherProp?: boolean;
+}
+```
+
+**Problem:** The `WidgeState` type is handwritten, requires repeating the types for slot props and any props that have defaults.
+
+```typescript
+export interface WidgetState extends WidgetProps {
+  icon?: ObjectShorthandProps<IconProps>;
+  text: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>>;
+  propWithDefaultValue: string;
+}
+```

```

Comment:
Why, indeed! :) The fact that it seems like it should be unnecessary is part of the motivation for this RFC.

It's necessary because:
* We need to convert the slots from `ShorthandProps` to `ObjectShorthandProps`
* The props with default values need to not be optional (`?`)
* We need to add the `ref` prop here. (In fact, I even missed doing that in the example.)

Reply:
Just chiming in that our `mergeProps` will accept shorthands of JSX, or object but once props are merged everything is actually an object

Reply:
@ling1726 Are you sure that `mergeProps` properly handles merging JSX shorthand? That's what `resolveShorthandProps` does, but I don't think `mergeProps` knows how to merge e.g. a JSX shorthand value with an ObjectShorthand value?

Reply:
@behowell ... yes my bad, you're right it's `resolveShorthandProps`, I have never seen it being used outside of `mergeProps` so for me the two go hand in hand... apologies for the confusion